src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorContentTypeChangeListener.cs

```diff
@@ -0,0 +1,184 @@
+    internal class RazorContentTypeChangeListener : ITextBufferContentTypeListener
+    {
+        private static readonly Guid HtmlLanguageServiceGuid = new Guid("9BBFD173-9770-47DC-B191-651B7FF493CD");
+
+        private readonly TrackingLSPDocumentManager _lspDocumentManager;
+        private readonly ITextDocumentFactoryService _textDocumentFactory;
+        private readonly LSPEditorFeatureDetector _lspEditorFeatureDetector;
+        private readonly LSPEditorService _editorService;
+        private readonly SVsServiceProvider _serviceProvider;
+        private readonly IEditorOptionsFactoryService _editorOptionsFactory;
+
+        [ImportingConstructor]
+        public RazorContentTypeChangeListener(
+            ITextDocumentFactoryService textDocumentFactory,
+            LSPDocumentManager lspDocumentManager,
+            LSPEditorFeatureDetector lspEditorFeatureDetector,
+            LSPEditorService editorService,
+            SVsServiceProvider serviceProvider,
+            IEditorOptionsFactoryService editorOptionsFactory)
+        {
+            if (textDocumentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textDocumentFactory));
+            }
+
+            if (lspDocumentManager is null)
+            {
+                throw new ArgumentNullException(nameof(lspDocumentManager));
+            }
+
+            if (lspEditorFeatureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(lspEditorFeatureDetector));
+            }
+
+            if (editorService is null)
+            {
+                throw new ArgumentNullException(nameof(editorService));
+            }
+
+            if (serviceProvider is null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (editorOptionsFactory is null)
+            {
+                throw new ArgumentNullException(nameof(editorOptionsFactory));
+            }
+
+            _lspDocumentManager = lspDocumentManager as TrackingLSPDocumentManager;
+
+            if (_lspDocumentManager is null)
+            {
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+                throw new ArgumentException("The LSP document manager should be of type " + typeof(TrackingLSPDocumentManager).FullName, nameof(_lspDocumentManager));
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+            }
+
+            _textDocumentFactory = textDocumentFactory;
+            _lspEditorFeatureDetector = lspEditorFeatureDetector;
+            _editorService = editorService;
+            _serviceProvider = serviceProvider;
+            _editorOptionsFactory = editorOptionsFactory;
+        }
+
+        public void ContentTypeChanged(ITextBuffer textBuffer, IContentType oldContentType, IContentType newContentType)
+        {
+            if (newContentType.IsOfType(RazorLSPConstants.RazorLSPContentTypeName))
```

Comment:
In theory you could have a content type change from RazorContentType1 to RazorContentType2 (in which case you don't want to do anything).
 this pattern:
   bool supportedBefore = ...
   bool supportedAfter = ...
   if (supportedBefore != supportedAfter) ...

seems safer,